rpcs3/Emu/Cell/SPURecompiler.cpp

```diff
@@ -6434,6 +6434,30 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 
 	void FSM(spu_opcode_t op)
 	{
+		// FSM following a comparison instruction
+		if (match_vr<s8[16], s16[8], s32[4]>(op.ra, [&](auto c, auto MP)
+		{
+			using VT = typename decltype(MP)::type;
+
+			if (auto [ok, x] = match_expr(c, sext<VT>(match<bool[std::extent_v<VT>]>())); ok)
+			{
+				if constexpr (std::extent_v<VT> == 4) // u32[4]
```

Comment:
Is this optimization for pshufd specifically? Or you want to avoid a bitcast?

Reply:
Yes, It's for pshufd.

Reply:
It doesn't emit on second path?

Reply:
It's not correct to use pshufd for the 16 and 8 bit compares, the intent of the pattern is to expand the result of one comparison to the whole register, if we use pshufd for 16/8 bit compares then we'll bring in stuff from the adjacent compares.

Reply:
We don't "use pshufd", it's LLVM optimizes shuffle instruction like this. The question is whether it'll optimize on second path with 12th byte or not. There is nothing incorrect about this.

Reply:
It doesn't emit pshufd on the second path.